internal/utils/gke.go

```diff
@@ -0,0 +1,508 @@
+	var operation *gkeapi.Operation
+	var err error
+
+	for i := 1; i < 12; i++ {
+		time.Sleep(time.Duration(i*i) * time.Second)
+		operation, err = svc.Projects.Locations.Clusters.Delete(ClusterRRN(config.Spec.ProjectID, config.Spec.Region, config.Spec.ClusterName)).Context(ctx).Do()
+		if err == nil {
+			return operation, nil
+		} else if !strings.Contains(err.Error(), "Please wait and try again once it is done") {
+			break
+		}
+	}
+	return operation, err
+}
+
+// GetClientset returns the clientset for a cluster
+func GetClientset(cluster *gkeapi.Cluster, ts oauth2.TokenSource) (kubernetes.Interface, error) {
+	capem, err := base64.StdEncoding.DecodeString(cluster.MasterAuth.ClusterCaCertificate)
+	if err != nil {
+		return nil, err
+	}
+	host := cluster.Endpoint
+	if !strings.HasPrefix(host, "https://") {
+		host = fmt.Sprintf("https://%s", host)
+	}
+
+	config := &rest.Config{
+		Host: host,
+		TLSClientConfig: rest.TLSClientConfig{
+			CAData: capem,
+		},
+		WrapTransport: func(rt http.RoundTripper) http.RoundTripper {
+			return &oauth2.Transport{
+				Source: ts,
+				Base:   rt,
+			}
+		},
+	}
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return clientset, nil
+}
+
+// GetGKEClient returns a gke client capable of making requests on behalf of the services account
+func GetGKEClient(ctx context.Context, secretsCache wranglerv1.SecretCache, config *gkev1.GKEClusterConfig) (*gkeapi.Service, error) {
+	ns, id := Parse(config.Spec.CredentialContent)
+	secret, err := secretsCache.Get(ns, id)
+
+	if err != nil {
+		return nil, err
+	}
+
+	dataBytes := secret.Data["gkeCredentialConfig-data"]
+
+	data := string(dataBytes)
+
+	return GetServiceClient(ctx, data)
+}
+
+// ValidateCreateRequest checks a config for the ability to generate a create request
+func ValidateCreateRequest(config *gkev1.GKEClusterConfig) error {
+	//TODO: check if these can even be nill/empty in a inported cluster.
+	if config.Spec.ProjectID == "" {
+		return fmt.Errorf("project ID is required")
+	} else if config.Spec.Zone == "" && config.Spec.Region == "" {
+		return fmt.Errorf("zone or region is required")
+	} else if config.Spec.Zone != "" && config.Spec.Region != "" {
+		return fmt.Errorf("only one of zone or region must be specified")
+	} else if config.Spec.ClusterName == "" {
+		return fmt.Errorf("cluster name is required")
+	}
+
+	for _, np := range config.Spec.NodePools {
+		if np.Autoscaling.Enabled != nil && *np.Autoscaling.Enabled {
+			if np.Autoscaling.MaxNodeCount == nil || np.Autoscaling.MinNodeCount == nil {
+				return fmt.Errorf("min and maxNodeCount can't be nil")
+			}
+			if *np.Autoscaling.MinNodeCount < 1 || *np.Autoscaling.MaxNodeCount < *np.Autoscaling.MinNodeCount {
+				return fmt.Errorf("minNodeCount in the NodePool must be >= 1 and <= maxNodeCount")
+			}
+		}
+	}
+
+	if config.Spec.IPAllocationPolicy == nil {
```

Comment:
Could this validation be performed in the validation functions?

Reply:
Not sure what you're asking for here, this is already in the `ValidateCreateRequest` function.